cmd/thanos/query.go

```diff
@@ -220,6 +220,8 @@ func registerQuery(app *extkingpin.App) {
 	tenantHeader := cmd.Flag("query.tenant-header", "HTTP header to determine tenant.").Default(tenancy.DefaultTenantHeader).String()
 	defaultTenant := cmd.Flag("query.default-tenant-id", "Default tenant ID to use if tenant header is not present").Default(tenancy.DefaultTenant).String()
 	tenantCertField := cmd.Flag("query.tenant-certificate-field", "Use TLS client's certificate field to determine tenant for write requests. Must be one of "+tenancy.CertificateFieldOrganization+", "+tenancy.CertificateFieldOrganizationalUnit+" or "+tenancy.CertificateFieldCommonName+". This setting will cause the query.tenant-header flag value to be ignored.").Default("").Enum("", tenancy.CertificateFieldOrganization, tenancy.CertificateFieldOrganizationalUnit, tenancy.CertificateFieldCommonName)
+	enforceTenancy := cmd.Flag("query.enable-tenancy", "Enable tenancy. Only responses where the value of the configured tenant-label-name and value of the tenant header matches are returned.").Default("false").Bool()
+	tenantLabel := cmd.Flag("query.tenant-label-name", "Label name to use when enforce tenancy when -querier.tenancy is enabled").Default(tenancy.DefaultTenantLabel).String()
```

Comment:
Do you mean `--query.enable-tenancy`? Maybe we could just enable tenancy if `--query.tenant-label-name` is not `""` or do you foresee other options regarding tenancy which would make it more sense to keep a separate boolean flag?

Reply:
> Do you mean `--query.enable-tenancy`?

I don't follow. Isn't that exactly what is there (except the dashes `--` which we don't add manually) ?

> Maybe we could just enable tenancy if `--query.tenant-label-name` is not `""` or do you foresee other options regarding tenancy which would make it more sense to keep a separate boolean flag?

Perhaps sometime later if we implement cross-tenancy there would need to be multiple options (disabled/single-tenant/cross-tenant), but not sure exactly how such an implementation would play out.

It would be nice to have one flag less, but I also think it's a little unclear from the flag name if `--query.tenant-label-name` enables tenancy enforcement. It also means, if you just want to use the defaults, you'd have to go and dig out corresponding default value on the receive side, and define it manually here, where's with this boolean flag, it's pretty easy to simply enable.

All minor details though, and I don't have a strong opinion, pros/cons with both. Mostly tried to make the implementation follow the approved proposal (as far as I can see there wasn't any specific discussion on these flags at the proposal stage). Any thoughts on this @douglascamata / @saswatamcode ?

Reply:
I think the argumentation regarding lack of clarity that a feature flag named `--query.tenant-label-name` triggers a lot more features to be turned on makes total sense. 

They are separate features:

* With `--query.tenant-label-name` I teach the query system how to identify a tenant. I get back logs and metrics with tenant information/dimension added to them. 

* Given I configured `--query.tenant-label-name`, I can turn on query tenancy enforcement (or not) to give more features on top of tenancy for the query path. 

I must confess that I also like the insight that in the future the `--query.enable-tenancy` flag might become an enum instead of a boolean. We could either model it today as an enum without too many changes, or we could leave it as boolean taking into consideration that making it an enum later might be a breaking change unless we add yet another flag (which I believe is not desired, we have plenty of flags already).

> It also means, if you just want to use the defaults, you'd have to go and dig out corresponding default value on the receive side, and define it manually here, where's with this boolean flag, it's pretty easy to simply enable.

+1 to this.

Reply:
> * With `--query.tenant-label-name` I teach the query system how to identify a tenant. I get back logs and metrics with tenant information/dimension added to them.

We don't currently use this information in any other place than for tenancy enforcement purposes. With https://github.com/thanos-io/thanos/pull/6794 we gather metrics about which tenant requests data, but we don't gather metrics about which tenants data is being accessed (i.e if enforcement is off tenant a could query tenant b's data). Perhaps that could be useful to do in the future though.

Reply:
A bit off-topic: kind of wishing all the config flags in any Thanos component could be loaded from a properly structured file (yaml, toml, json, whatever). 😅